### PR TITLE
fix(tab):generic id type added to TabItem

### PR DESCRIPTION
**Pull request by tatata96**
**Related Issue;**
- #172

### CODE REVIEW

**Review by gulcinuras — COMMENTED** [review]
@tatata96 I asked a small change, but we can also discuss if that makes sense or not 😸 
Also, I noticed that in `12-Tab.stories` we are not providing the `initialState` in a `Record` type, so that it gives a type error.

Therefore, I think we can update the `initialState` on line 39 like this:
```
<StateProvider initialState={{index: 0}}>
```
and update the other places we use that state.

File: src/tab/Tab.tsx [hunk 1]
```diff
@@ -6,15 +6,15 @@ import classNames from "classnames";
...
+  id: ID;
   content: React.ReactNode;
   icon?: React.ReactNode;
   isDisabled?: boolean;
 };
 
-interface UncontrolledTabProps {
-  items: TabItem[];
+interface UncontrolledTabProps<ID> {
+  items: TabItem<ID>[];
```
[the remark this answers is not in the document]
Comment (ErenYLCN) [reply src/tab/Tab.tsx:17]: I think we can discuss this with the team

**Review by ErenYLCN — APPROVED** [review]

File: src/tab/Tab.tsx [hunk 2]
```diff
@@ -6,15 +6,15 @@ import classNames from "classnames";
 import TabHeaderItem from "./header/item/TabHeaderItem";
 import List from "../list/List";
 
-export type TabItem = {
-  id: string;
+export type TabItem<ID = string | number> = {
```
Comment (Anlerkan) [inline src/tab/Tab.tsx:9]: In this version, we can pass boolean or false to the generic type. We can even pass array :D Should we? IMO, we should be able to pass string or number only.
```suggestion
export type TabItem<ID extends string | number = string | number> = {
```

Comment (tatata96) [reply src/tab/Tab.tsx:9]: You're right. I missed it and gave default types only, I should have limited it using constraints. 
Thank you for noticing ✌️

**Review by gulcinuras — APPROVED** [review]

**Review by edizcelik — APPROVED** [review]

File: src/tab/Tab.tsx [hunk 3]
```diff
@@ -6,15 +6,15 @@ import classNames from "classnames";
 import TabHeaderItem from "./header/item/TabHeaderItem";
 import List from "../list/List";
 
-export type TabItem = {
-  id: string;
+export type TabItem<ID extends string | number = string | number> = {
```
Comment (yasincaliskan) [inline src/tab/Tab.tsx:9]: Just wonder, if the ID can be a string or number, should we extend it with these types? Wouldn't be okay like this? 🤔 
```suggestion
export type TabItem<ID = string | number> = {
```

**Review by yasincaliskan — COMMENTED** [review]
🎉

File: src/tab/Tab.tsx (hunk 3 above)
Comment (tatata96) [reply src/tab/Tab.tsx:9]: With the current version the type of ID can only string or number (same as your suggestion), but if you don't specify a type argument, the default type will be one of those two.

File: stories/12-Tab.stories.tsx [hunk 4]
```diff
@@ -45,21 +45,21 @@ storiesOf("Tab", module).add("Tab", () => (
...
           </div>
 
           <Tab
             items={tabItems}
-            activeTabIndex={state}
+            activeTabIndex={state.index}
             onTabChange={(index) => {
               console.log("tab changed to index: ", index);
-              setState(index);
+              setState({index: index});
```
Comment (Anlerkan) [inline stories/12-Tab.stories.tsx:62]: I think we can use shorthand version
```suggestion
              setState({index});
```

File: src/tab/header/item/TabHeaderItem.tsx [hunk 5]
```diff
@@ -7,7 +7,7 @@ import {TabItem} from "../../Tab";
 import ListItem from "../../../list/item/ListItem";
 
 type TabHeaderItemProps = {
-  tab: TabItem;
+  tab: TabItem<string | number>;
```
Comment (Anlerkan) [inline src/tab/header/item/TabHeaderItem.tsx:10]: Shouldn't we get the ID shape from `TabHeaderItemProps` ?

File: src/tab/Tab.tsx [hunk 6]
```diff
@@ -26,16 +26,16 @@ interface UncontrolledTabProps {
...
   | {
-      activeTabIndex?: number;
-      onTabChange?: (index: number) => void;
-    };
+    activeTabIndex?: number;
+    onTabChange?: (index: number) => void;
+  };
 
-export type TabProps = ControlledTabProps & UncontrolledTabProps;
+export type TabProps = ControlledTabProps & UncontrolledTabProps<string | number>;
```
Comment (Anlerkan) [inline src/tab/Tab.tsx:38]: Also here, should not we derive the id shape from `TabProps`?

**Review by Anlerkan — CHANGES_REQUESTED** [review]

**Review by yasincaliskan — APPROVED** [review]

**Review by jamcry — APPROVED** [review]

**Pull request merged**